autorest/codegen/__init__.py

```diff
@@ -141,12 +143,60 @@ def _build_package_dependency() -> Dict[str, str]:
             "dependency_msrest": "msrest>=0.6.21",
         }
 
+    @staticmethod
+    def _build_with_security_definition(yaml_data: Dict[str, Any], credential_info: Dict[Any, Any]):
+        security_yaml = yaml_data.get("security", {})
+        if security_yaml.get("authenticationRequired"):
+            # the key word is "AADToken" and "AzureKey"
+            credential_scopes = set()
+            key_header_name = ""
+            for scheme in security_yaml.get("schemes"):
+                if CredentialInfo.AAD_TOKEN == scheme["type"]:
+                    credential_scopes.update(scheme["scopes"])
+                elif CredentialInfo.AZURE_KEY == scheme["type"]:
+                    # only accept the last one
+                    key_header_name = scheme["headerName"]
+            if credential_scopes:
+                credential_info[CredentialInfo.SCOPES] = list(credential_scopes)
+            if key_header_name:
+                credential_info[CredentialInfo.KEY_HEADER_NAME] = key_header_name
+
+        if CredentialInfo.SCOPES in credential_info:
```

Comment:
I think this check should be changed to `if credential_info['CredentialInfo.SCOPES']`, bc when cleaning up the code, scopes is always in credential info. Same for the below.

Reply:
we should have tests that would catch this. To make sure all scenarios are working correctly, I think we should also try the following generations (still just using your two swaggers SecurityAAD and SecurityKey). I wouldn't worry too much about the cases where swagger + flags are in agreement, we want to test the disagreements.

| Swagger   | Flags | Output |
| -------------------| ----------- |----------- |
| `SecurityAAD`  |  `--add-credential` and `--credential-default-policy-type=AzureKeyCredentialPolicy`    |       `AzureKey` auth      |
| `SecurityAAD`  |  `--azure-arm`    |   `ARMChallengeAuth` (not `BearerTokenAuth`)      |
| `SecurityKey`  |  `--add-credential`   |       `BearerTokenAuth`, bc that's the default policy when users pass in `--add-credential` and not `azure-arm`     |
| `SecurityKey`  |  `--azure-arm`   |   `ARMChallengeAuth`     |

Reply:
it would be great to also keep the generation of these limited to the `tasks.py` file. See [this](https://github.com/Azure/autorest.python/blob/autorestv3/tasks.py#L556) for inspiration, where we pass in `override_flags`

Reply:
> I think this check should be changed to `if credential_info['CredentialInfo.SCOPES']`, bc when cleaning up the code, scopes is always in credential info. Same for the below.

`credential_info` maybe null dict and `if credential_info[CredentialInfo.SCOPES]` will raise exception

Reply:
Will update test after refactor.

Reply:
updated: I commit the refactor code. Now I am working on the test

Reply:
@iscai-msft , updated: add `regenerate_security`